### PR TITLE
Fix `mut` static task queue in SGX target

### PR DESCRIPTION
**Pull request by raoulstrackx**
[PR 125046](https://github.com/rust-lang/rust/pull/125046) prevents mutable references to statics with `#[linkage]`. Such a construct was used with the tests for the `x86_64-fortanix-unknown-sgx` target. This PR fixes this and cleans up code a bit in 5 steps. Each step passes CI:

- The `mut` static is removed, and `Task` explicitly implements `Send`
- Renaming of the `task_queue::lock` function
- Pass function for `Thread` as `Send` to `Thread::imp` and update when `Packet<'scope, T>` implements `Sync`
- Storing `Task::p` as a type that implements `Send`
- Letting the compiler auto implement `Send` for `Task`

cc: @jethrogb

### CODE REVIEW

**Comment by bjorn3** [comment]
I think in https://github.com/rust-lang/rust/pull/125046 I should have only denied `#[linkage] static mut` for extern statics.

**Comment by raoulstrackx** [comment]
> I think in #125046 I should have only denied `#[linkage] static mut` for extern statics. Basically moving the code I added in compiler/rustc_codegen_ssa/src/codegen_attrs.rs into the `if tcx.is_foreign_item(did)` arm.

Right, but even then I don't see why `Task::p` wouldn't need to implement `Send` as it's still passed from one thread (added to the list in `Thread::new`) to another (removed from the list in `Thread::entry`).

Anyway, I believe this PR still cleans up the code.

**Comment by workingjubilee** [comment]
Yeah, seems preferable this way.
@bors r+

File: library/std/src/sys/pal/sgx/thread.rs [hunk 1]
```diff
@@ -45,18 +45,12 @@ mod task_queue {
...
     #[export_name = "_ZN16__rust_internals3std3sys3sgx6thread10TASK_QUEUEE"]
-    static mut TASK_QUEUE: Option<Mutex<Vec<Task>>> = None;
+    static TASK_QUEUE: Mutex<Vec<Task>> = Mutex::new(Vec::new());
 
-    pub(super) fn lock() -> MutexGuard<'static, Vec<Task>> {
-        unsafe {
-            TASK_QUEUE_INIT.call_once(|| TASK_QUEUE = Some(Default::default()));
-            TASK_QUEUE.as_ref().unwrap().lock().unwrap()
-        }
+    pub(super) fn tasks() -> MutexGuard<'static, Vec<Task>> {
```
Comment (jethrogb) [inline library/std/src/sys/pal/sgx/thread.rs:52]: Why change the name of the function? It's not clear anymore in usage sites that this is returning a mutex guard, which complicates lock analysis.

Comment (raoulstrackx) [reply library/std/src/sys/pal/sgx/thread.rs]: The reasoning was that `lock` doesn't tell the developer much more than it's locking something. Looking at it again, that's not really an issue here as we always explicitly mention the module name as well. I've removed the commit that renamed the function.

**Comment by workingjubilee** [comment]
Waiting for @jethrogb's approval.

**Comment by workingjubilee** [comment]
hm in fact
@bors delegate=jethrogb

**Comment by fmease** [comment]
bors sleepy @bors r-

**Comment by jethrogb** [comment]
@fmease I'm confused about what you're trying to do. This PR is merged into master.

**Comment by fmease** [comment]
Just maintenance. Sometimes @bors's [queue](https://bors.rust-lang.org/queue/rust) gets out of sync with GitHub (e.g., containing already merged PRs). While I could press the "synchronize" button, it's a bit destructive and buggy (try-builds get upgraded to builds, disapproved PRs suddenly show up in the queue, etc.). Therefore it's typical for someone to swoop in and manually "sync" the queue by r-'ing already merged PRs.

**Pull request merged**